### PR TITLE
Style mixins

### LINKED ISSUE #2: Final default styles and theme mixins/custom props

### PR DESCRIPTION
**Pull request by manolo**
Fixes #2

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-context-menu/8)

### CODE REVIEW

File: demo/advanced.html [hunk 1]
```diff
@@ -25,11 +25,41 @@
...
     <demo-snippet>
       <template>
         <grid-example></grid-example>
         <dom-module id="grid-example">
+          <style>
+            :host {
+              --vaadin-context-menu-overlay: {
+                box-shadow: none;
+                border: 1px solid var(--primary-color);
+                width: 110px;
```
Comment (jouni) [inline demo/advanced.html]: Maybe min-width instead?

File: demo/advanced.html [hunk 2]
```diff
@@ -25,11 +25,41 @@
...
+                padding: 10px 0;
+              };
+              --vaadin-context-menu-item: {
+                min-height: 24px;
+                padding: 0 10px;
+              };
+              --paper-menu-selected-item: {
+                font-weight: normal;
+              };
+              --paper-menu-background-color: #e1f3fc;
```
Comment (jouni) [inline demo/advanced.html]: Where does this color come from?

File: vaadin-context-menu.html [hunk 3]
```diff
@@ -2,26 +2,44 @@
...
+        padding: 16px 0;
+        background: var(--paper-menu-background-color, --primary-background-color);
         @apply(--shadow-elevation-6dp);
+        @apply(--vaadin-context-menu-overlay);
       }
 
       :host ::content paper-menu {
-        width: calc(64px * 4);
-        padding: 16px 0;
+        padding: 0px;
```
Comment (jouni) [inline vaadin-context-menu.html]: Unnecessary `px`

**Comment by manolo** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[vaadin-context-menu.html, line 35 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRXlnR75q_7yGp1gA_f-r1-35:-KRbGRA-DB4FkSItQbNW:b-896fix) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/1657c668ff332da6734b34ee6fb6c41540ae94d4/vaadin-context-menu.html#L35)):*
<details><summary><i>Previously, jouni (Jouni Koivuviita) wrote…</i></summary>
> Unnecessary `px`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8)*

**Comment by tomivirkki** [comment]
*[vaadin-context-menu.html, line 21 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRbMR64nKs5H-gxPQjG:-KRbMR64nKs5H-gxPQjH:brbnepp) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/5fb53d8befb10796130ce4b29a63d90ce9c77dd4/vaadin-context-menu.html#L21)):*
> ```HTML
> <dom-module id="vaadin-context-menu">
>   <template>
>     <style is="custom-style">
> ```

This shouldn't be needed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8)*

**Comment by tomivirkki** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[vaadin-context-menu.html, line 27 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRbSGmgVKLmNNSBgnsr:-KRbSGmgVKLmNNSBgnss:bgend59) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/5fb53d8befb10796130ce4b29a63d90ce9c77dd4/vaadin-context-menu.html#L27)):*
> ```HTML
> 
>       .dropdown-content {
>         width: calc(64px * 4);
> ```

Why use calc and not just 256px?

---

*[vaadin-context-menu.html, line 42 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRbRO3KRrUXOaOkKMbp:-KRbRO3KRrUXOaOkKMbq:bvn6jgk) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/5fb53d8befb10796130ce4b29a63d90ce9c77dd4/vaadin-context-menu.html#L42)):*
> ```HTML
>         min-height: 32px;
>         padding: 0 24px;
>         @apply(--vaadin-context-menu-item);
> ```

This should target the paper-item only (as documented). Now the selector also targets paper-item-body which is unexpected. I'd completely remove paper-item-body from the selector as it adds too much control over the item content.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8)*

**Comment by manolo** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[vaadin-context-menu.html, line 27 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRbSGmgVKLmNNSBgnsr:-KRbWSHbLBK3_Wm86VJh:biby127) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/5fb53d8befb10796130ce4b29a63d90ce9c77dd4/vaadin-context-menu.html#L27)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary>
> Why use calc and not just 256px?
</details>
To know that it is 4 times 64 which is the material design specification.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8)*

**Comment by manolo** [comment]
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[vaadin-context-menu.html, line 21 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRbMR64nKs5H-gxPQjG:-KRbZGaghf_-OV7j2ZDq:b-896fix) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/5fb53d8befb10796130ce4b29a63d90ce9c77dd4/vaadin-context-menu.html#L21)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary>
> This shouldn't be needed
</details>
Done.

---

*[vaadin-context-menu.html, line 42 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRbRO3KRrUXOaOkKMbp:-KRbZDaYBD0gJ6WXlEpE:b-896fix) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/5fb53d8befb10796130ce4b29a63d90ce9c77dd4/vaadin-context-menu.html#L42)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary>
> This should target the paper-item only (as documented). Now the selector also targets paper-item-body which is unexpected. I'd completely remove paper-item-body from the selector as it adds too much control over the item content.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8)*

**Comment by manolo** [comment]
Review status: 2 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[demo/advanced.html, line 28 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRb_hq7aonnD81DKteE:-KRbbPigYBLMBVhbvqmD:b-896fix) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/ae1cf8fe17b9273f19cb5f33a0d855e9f9b93c0a/demo/advanced.html#L28)):*
<details><summary><i>Previously, tehapo (Teemu Pöntelin) wrote…</i></summary>
> Polymer should be written with an uppercase "P".
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8)*

**Comment by tomivirkki** [comment]
*[vaadin-context-menu.html, line 29 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRbmA1C6dElv3sRti8p:-KRbmA1DNP-z2b0rHTKy:b-na0l0c) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/40706e811b6e8fc85217507079df1772a93d509a/vaadin-context-menu.html#L29)):*
> ```HTML
>         width: calc(64px * 4);
>         padding: 16px 0;
>         background: var(--paper-menu-background-color, --primary-background-color);
> ```

Styling the content wrapper with paper-menu specific styles makes sense only when there's a paper-menu inside. I't might be cleaner to remove the padding and background from `.dropdown-content`completely and move the `padding 16px 0` to `:host ::content paper-menu` styles

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8)*

**Comment by manolo** [comment]
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[vaadin-context-menu.html, line 29 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRbmA1C6dElv3sRti8p:-KRbxdCjoM52ZfohqUZ2:b-iyb0hc) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/40706e811b6e8fc85217507079df1772a93d509a/vaadin-context-menu.html#L29)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary>
> Styling the content wrapper with paper-menu specific styles makes sense only when there's a paper-menu inside. I't might be cleaner to remove the padding and background from `.dropdown-content`completely and move the `padding 16px 0` to `:host ::content paper-menu` styles
</details>
My idea is that the user can customise completely the dropdown content. We, by default, have to set styles here, and to me makes sense to try to reuse paper colors, shadows, and paper menu, but we actually don't depend on it.  When the user needs a different design he has the --vaadin-context-menu-overlay mixin to override our defaults. The user still can use paper-menu and paper-item mixins if he uses those components.
If we style in the paper-menu we need another mixin for the component like --vaadin-context-menu-paper-menu so as he can override our defaults, which does not sounds good to me. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8)*

**Comment by tomivirkki** [comment]
*[vaadin-context-menu.html, line 29 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRbmA1C6dElv3sRti8p:-KRc4mR0o8fKH7LvwSaz:b-xez6hb) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/40706e811b6e8fc85217507079df1772a93d509a/vaadin-context-menu.html#L29)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary>
> My idea is that the user can customise completely the dropdown content. We, by default, have to set styles here, and to me makes sense to try to reuse paper colors, shadows, and paper menu, but we actually don't depend on it.  When the user needs a different design he has the --vaadin-context-menu-overlay mixin to override our defaults. The user still can use paper-menu and paper-item mixins if he uses those components.
> If we style in the paper-menu we need another mixin for the component like --vaadin-context-menu-paper-menu so as he can override our defaults, which does not sounds good to me. 
</details>
"My idea is that the user can customise completely the dropdown content" > Exactly, that's why the dropdown-content itself should by default be merely a container for any element. Now it derives background color from a specific paper component (paper-menu).

The padding is also an unneeded default: The component that's placed inside can have a padding of it's own. The current approach (in case of a paper-menu) affects that by removing the default padding from the paper-menu and adding padding to the paper-content. Clearly the --paper-menu-background-color is then used to try to cover the transparent background peeking from behind because of this. That's 3 CSS rules, where you really should only need 1.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8)*

**Comment by tomivirkki** [comment]
*[vaadin-context-menu.html, line 36 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRcdOzmGugbFagvDJHh:-KRcdOzn8vd768ARZBkZ:budzgw6) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/246167b3d6a7366826ccf87817f6497f4cb823a4/vaadin-context-menu.html#L36)):*
> ```HTML
>       :host ::content paper-menu {
>         padding: 16px 0;
>         @apply(--vaadin-context-paper-menu);
> ```

paper-menu and paper-item don't need mixins because the developer has direct access to them anyway.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8)*

**Comment by manolo** [comment]
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[vaadin-context-menu.html, line 29 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRbmA1C6dElv3sRti8p:-KRcSjKu7jAUT4GUZpzz:b-51hwll) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/40706e811b6e8fc85217507079df1772a93d509a/vaadin-context-menu.html#L29)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary>
> "My idea is that the user can customise completely the dropdown content" > Exactly, that's why the dropdown-content itself should by default be merely a container for any element. Now it derives background color from a specific paper component (paper-menu).
> 
> The padding is also an unneeded default: The component that's placed inside can have a padding of it's own. The current approach (in case of a paper-menu) affects that by removing the default padding from the paper-menu and adding padding to the paper-content. Clearly the --paper-menu-background-color is then used to try to cover the transparent background peeking from behind because of this. That's 3 CSS rules, where you really should only need 1.
</details>
We have a compromise of giving a ready to use material design component to the user. Paper-menu and paper-item do not accomplish cascading menu specification

---

*[vaadin-context-menu.html, line 36 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRcdOzmGugbFagvDJHh:-KRdupeBkvKw5oqotoLc:b1zowib) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/246167b3d6a7366826ccf87817f6497f4cb823a4/vaadin-context-menu.html#L36)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary>
> paper-menu and paper-item don't need mixins because the developer has direct access to them anyway.
</details>
We should provide a vaadin-context-menu that fulfil MD specification by default (https://material.google.com/components/menus.html), we should not make the user style every single vaadin-context-menu that they use in their app.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8)*

**Comment by tomivirkki** [comment]
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[vaadin-context-menu.html, line 29 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRbmA1C6dElv3sRti8p:-KRgvrNIZ8QS-N7i7vQ0:b-o5p33g) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/40706e811b6e8fc85217507079df1772a93d509a/vaadin-context-menu.html#L29)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary>
> We have a compromise of giving a ready to use material design component to the user. Paper-menu and paper-item do not accomplish cascading menu specification
</details>
`paper-menu` isn't cascading. It's MD the way a non-cascading menu should be. We'll anyway have to provide a cascading menu component in the future, but a user should also be able to place a `paper-menu` inside a `vaadin-context-menu` and it should have default appearance for a non-cascading menu.

---

*[vaadin-context-menu.html, line 36 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRcdOzmGugbFagvDJHh:-KRgw7G5oOI8O_jwhVeX:b-rgmzim) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/246167b3d6a7366826ccf87817f6497f4cb823a4/vaadin-context-menu.html#L36)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary>
> We should provide a vaadin-context-menu that fulfil MD specification by default (https://material.google.com/components/menus.html), we should not make the user style every single vaadin-context-menu that they use in their app.
</details>
This note was merely about not providing unnecessary mixins to style elements in the light DOM. It looks ok now.

---

*[vaadin-context-menu.html, line 28 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRgwRoE5PXwmfMXmMdM:-KRgwRoE5PXwmfMXmMdN:b-x7feft) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/795336f22e0872b9daea6ee7d260332e9b563ff6/vaadin-context-menu.html#L28)):*
> ```HTML
>         width: calc(64px * 4);
>         background: var(--primary-background-color, #ffffff);
>         padding: 8px 0;
> ```

This padding should still not be here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8)*

**Comment by manolo** [comment]
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[vaadin-context-menu.html, line 29 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRbmA1C6dElv3sRti8p:-KRgyg363kt1XsfLnQE0:bvvs8g3) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/40706e811b6e8fc85217507079df1772a93d509a/vaadin-context-menu.html#L29)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary>
> `paper-menu` isn't cascading. It's MD the way a non-cascading menu should be. We'll anyway have to provide a cascading menu component in the future, but a user should also be able to place a `paper-menu` inside a `vaadin-context-menu` and it should have default appearance for a non-cascading menu.
</details>
I know!, but in the spec they give the dimensions for what they show as contextual menus in the cascading section, that is the section I was refering to with my comment. 

---

*[vaadin-context-menu.html, line 28 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRgwRoE5PXwmfMXmMdM:-KRh-gczo6NVz1_RYibg:b-896fix) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/795336f22e0872b9daea6ee7d260332e9b563ff6/vaadin-context-menu.html#L28)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary>
> This padding should still not be here
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8)*

**Comment by Saulis** [comment]
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[demo/index.html, line 66 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRh-rUEjkpxTMSXw_en:-KRh-rUEjkpxTMSXw_eo:b-m8qrzr) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/3bb5bdf7292b65a4c488b0ffd1adb1db9238733e/demo/index.html#L66)):*
> ```HTML
>       <template>
>         <dom-module id="context-example2">
>           <style>
> ```

No need to wrap the styles in a `<dom-module>`, just give the menu an `id` to prevent styles from leaking to another `<demo-snippet>` 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8)*

**Comment by manolo** [comment]
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[demo/index.html, line 66 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-KRh-rUEjkpxTMSXw_en:-KRh3lga771BPpRJ8WdZ:byepdc0) ([raw file](https://github.com/vaadin/vaadin-context-menu/blob/3bb5bdf7292b65a4c488b0ffd1adb1db9238733e/demo/index.html#L66)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary>
> No need to wrap the styles in a `<dom-module>`, just give the menu an `id` to prevent styles from leaking to another `<demo-snippet>` 
</details>
It was in this way in order to add some action on selection.
Also the following seems not working. Tested style block in the template, and with/without the is=custom-style.

But anyway I have removed this example.

```
<demo-snippet>
  <style is="custom-style">
    #mmenu {
       --paper-menu: {
         ...
       };
       ...
    }
  </style>
  <template>
        <button id="opener">right-click me</button>
        <vaadin-context-menu target-selector="#opener" id="mmenu">
          <paper-menu>
           ....
          </paper-menu>
        </vaadin-context-menu>
  </template>
</demo-snippet>

```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8)*

**Comment by tomivirkki** [comment]
:lgtm:

---

Reviewed 2 of 4 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-context-menu/8#-:-KRhHQlMr8JyKP2GUdkt:bnfp4nl)*

**Pull request merged**